buildbot: Add support for creating Steam builds and uploading them to SteamPipe

OatmealDome opened the pull request:
Supersedes #159.

This PR adds builders for Steam builds (release and PR). Release builders will automatically upload their artifacts to SteamPipe and set them as live on a beta branch. PR builders will upload their artifacts to `dl.dolphin-emu.org` as normal.

Here's how the release builder works:

* The `release-steam` builder gets run along with the usual release builders.
* This builder then starts the `release-steam-all` triggerable, which in turn starts the `release-steam-*`builders and waits for them to finish.
* The final builds are uploaded to the buildbot master at `/tmp/steam/<commit hash>/win.7z`, `mac.dmg`, and `lin.7z`.
* `release-steam` extracts all the builds to a `content` subdirectory.
* The `ContentBuilder` toolset from the Steamworks SDK is then used to upload the builds to SteamPipe using the `steampipe_app_build.vdf` app build script.

PR builders are just ran directly from the standard `pr` `AnyBranchScheduler`.

Prerequisites:

* The Steamworks SDK ([download here](https://partner.steamgames.com/downloads/steamworks_sdk_155.zip), login required) needs to be installed on the buildbot master somewhere.
* [Valve recommends that a separate, dedicated build account be used to upload builds](https://partner.steamgames.com/doc/sdk/uploading#Build_Account). The shell script `upload_to_steampipe.sh` needs to be updated with the account's username.
* Once an account is created, the Steamworks SDK requires an initial login to complete a Steam Guard challenge (if necessary) and create cached credentials. This can be done with the following:
`/path/to/sdk/tools/ContentBuilder/builder_linux/steamcmd.sh +login <username> <password>`

Other notes:

* While it is possible to automatically set builds live on branches ("betas" in user-facing terminology), it is not possible to automatically set builds live on the default (non-beta) branch. For that, a human needs to manually set builds live on the SteamPipe dashboard.
* `steamcmd`'s cached credentials can expire, but I can't find any documentation on when / how this occurs. (I find that my cached credentials become invalidated after some time, but this is after months of not using them. Not sure if this happens because of them being unused or if they're just on a set timer.) It might be necessary to somehow pass the account's password to the shell script so it can be provided to `steamcmd` on every upload for authentication.
* I tried to test as much of this PR as possible, but there are some things I was unable to test, like all the "upload to `dl.dolphin-emu.org` stuff.

delroth commented:
Do we actually want every build on SteamPipe or just every beta? My intuition would have been the latter, just like we do for Android/Play Store.

OatmealDome commented:
I was thinking it might be convenient for users to be able to switch to the "dev" beta for testing development builds, but I have no strong opinions either way.

How does the Play Store release system work?

delroth commented:
We have an hourly cron polling the beta branch and uploading the latest beta to the Play Store if it's not already up to date. Simpler approach imo because we can have the buildbot just generate artifacts, and everything related to SteamPipe upload can be done in an external script.

https://github.com/dolphin-emu/sadm/blob/nix/roles/android-updater/updater.py

delroth commented:
> The Steamworks SDK ([download here](https://partner.steamgames.com/downloads/steamworks_sdk_155.zip), login required) needs to be installed on the buildbot master somewhere.

This will need a workaround, because I can't add a source that is not openly fetchable to our infra configuration, and I'm definitely not installing anything manually and creating more technical debt for myself in the future.

delroth commented on buildbot/master.cfg:
```diff
@@ -549,6 +612,86 @@ def make_dolphin_linux_build(mode="normal"):
+    pr = "pr" in mode
+
+    f.addStep(GitNoBranch(repourl="https://github.com/dolphin-emu/dolphin.git",
+                          progress=True, mode="incremental", workdir="build/qt"))
```
That workdir seems wrong.

OatmealDome replied:
Would `build/build/qt` work? Was worried putting it into the CMake build directory would mess something up.

OatmealDome commented on buildbot/steampipe_app_build.vdf:
```diff
@@ -0,0 +1,48 @@
+"AppBuild"
```
I was thinking it should be here as a buildbot-related thing since it relies on the folder layout of the extracted 7z + dmg files. (For example, [the `Dolphin-x64` subdir is named by the buildbot during 7z compression](https://github.com/dolphin-emu/sadm/blob/9da3d33905554e3eaa8bbab7c067df2f71827edf/buildbot/master.cfg#L188).)

delroth replied:
But things like which binaries should/shouldn't be included is really a property of the build itself, no?

Could we somehow make this VDF file more independent of buildbot implementation details if we have the .sh build script do a bit more massaging with the extraction of the archives?

I agree it's kind of borderline though... Not sure one way or another.

OatmealDome replied:
> Could we somehow make this VDF file more independent of buildbot implementation details if we have the .sh build script do a bit more massaging with the extraction of the archives?

Oh, definitely. Just need to make sure everything is in `win`, `mac`, and `lin` directly and get rid of everything unnecessary that we exclude using `FileExclusions`. Then we can have something like this:

```
{
    "AppID" "1941680"
    "Desc" "DOLPHIN_BUILD_NUMBER"
    "SetLive" "dev-sniper"

    "ContentRoot" "content"
    "BuildOutput" "output"

    "Depots"
    {
        "1941682" // Windows
        {
            "FileMapping"
            {
                "LocalPath" "win/*"
                "DepotPath" "."
                "Recursive" "1"
            }
        }
        "1941683" // macOS
        {
            "FileMapping"
            {
                "LocalPath" "mac/*"
                "DepotPath" "."
                "recursive" "1"
            }
        }
        "1941684" // Linux
        {
            "FileMapping"
            {
                "LocalPath" "lin/*"
                "DepotPath" "."
                "Recursive" "1"
            }
        }
    }
}
```

OatmealDome commented:
OK, rebased + made some changes.

I made `steampipe_app_build.vdf` more generic so it can be PR'd into the main repo if needed. The directory structure is now set up by the shell script.

I also added a line to the shell script which loads the Steam build account password from a file. The placeholder needs to be replaced when the actual path is known.

Also, I fixed the workdir issue. Was a mistake from an earlier rebase.

delroth commented:
Can you retarget this towards the pre-nix branch and reopen?
